README.md
Based on the original Tektronix 4051 Javascript Emulator by Dave R. found here:

https://drive.google.com/drive/folders/1uXHa34qDizMKOGZWemFbpUet1JlYxOEx

and enhanced by Jon Stanley found here:

https://github.com/jonbstanley/Tek405xEmulator

---

For details of releases including bug fixes and what is new see <a href="https://github.com/Twilight-Logic/Tek405xEmulator/blob/master/Releases.md">Releases.md</a>.

---

<B>Additions by Monty McGraw:</B>

Update 02 Nov 2021

  - the display has been updated to enable refresh graphics

Previous:

 - Added new "UNIVERSAL" HTML emulator file that is compatible with all 4050 Control Characters

I plan to begin converting ALL my older program captures - to a universal format compatible with the universal emulator HTLM file AND the 4051, 4052, 4052A and 4054A computers.

For the older programs from the Programs directory, and my older programs on the 4050 Program repository that have Character, backspace, underline - execute the older HTML file

---

Detailed description by Dave R.:


Tektronix 4051 Javascript emulator
==================================

This is a partially complete emulation of the Tektronix 4051 computer.

The emulation consists of a 6800 CPU complete with the 'hardware' and 
original ROM images. No internal tape drive is provided. Loading the
emulator with a BASIC program is accomplished through the browser interface
and the use of an emulated GPIB device (#1). There is also a Real Time Clock
configured as GPIB device #2.

Some of the keyboard keys are a bit 'wonky' at this point in time (!) as is
some of the screen display (for example the blinking cursor). Consider this
release as a pre-Alpha... The emulator is also considerably faster than the
original Tektronix 4051 machine. This is generally good news - except when
trying to pilot the Lunar Lander down (where this is virtually impossible)!

This emulator was written on an Apple iMac with the Safari browser. The emulator
requires a browser that is HTML 5 compliant (as some of the features used for
loading the BASIC program into the emulator). I have saved the source files in
such a manner that they should be readable and modifiable on a Windows platform - 
although I have not actually tested this out.

To run the emulator - double-click on the 'jsTEKTRONIX4051.html' icon. Your browser
should start and you should be presented with a large (1024 by 1024) black screen.
Scroll down a bit and you should find a button marked [start]. Click the [start] 
button. A little bit of diagnostic text (DER: MC6800.js >>>>> CRTRST: <<<<<) should 
appear in the debug log window below the button.

You should also see a blue square in the top left-hand corner of the screen. This is
(or will be when I have finished it!) the flashing green cursor...

Congratulations, the emulator seems to be running...

To load a BASIC program from the host operating system, the following procedure should
be followed:

Scroll down a bit to the buttons again.

Click the [Choose File] button and navigate to the directory containing the LINES.txt
file. Select this file. A message should appear in the debug log window as follows
"Selected file loaded from host operating system.".

In the Tektronix window type the command 'OLD @1:' followed by the <RETURN> key. The
selected BASIC program should be loaded into the emulator and the blue cursor should 
reappear on the next line. Type RUN and you should be away...

If you have tried the above procedure without reading to here - you will have identified
a problem with the keyboard. I have tried my best to map the Tektronix 4051 key matrix
to a 'standard' PC keyboard - but failed dismally in a number of areas! This is also not
helped by the shifted characters on a 'standard' keyboard being different to those on the
Tektronix keyboard either! I have mapped the keys as follows (again, noting that these
are off my iMac keyboard and not a PC). Also, I have a wired keyboard on my iMac and not
the wireless one - so I have access to the numeric pad.

Tektronix @         = Keyboard `
Tektronix :         = Keyboard '
Tektronix PAGE      = Function key 13.
Tektronix RUBOUT    = <DELETE> key.
Tektronix BREAK     = Function key 14.
Tektronix CLEAR     = Function key 15.
Tektronix AUTO NUM  = Function key 16.
Tektronix STEP PROG = Function Key 17.

If you hate me for this mapping (!) you have the source code so you can change it
yourself. Look for the function this.HandleKeyboardEvent in the file TEKTRONIX4051.js.

The CAPS LOCK, Control and Shift keys work - albeit with some limitations on the
characters that are actually displayed! I would suggest typing the characters on your 
keyboard and seeing what appears on the emulated Tektronix screen as a start.

You can't edit programs on the Tektronix emulation and save them back to the PC's disk.
Standard HTML and Javascript prohibits this. My expectation is that people will have
(or edit) the Tektronix BASIC programs using their favourite text editor on the PC and
download the program to run on the emulator.

I hope you find this emulator of interest. If you do, have identified bugs or have
helpful suggestions for improvement then please contact me at: 
daver21145 AT gmail DOT com - with text / character conversions as appropriate.

Regards,

Dave


Releases.md
<B>Updates by Twilight Logic</B>

Update 27 Jul 2022
  - bug fix: fixed problem with mute button not working

Update 22 Jul 2022
  - bug fix: fixed isse where the index was not getting cleared after a DelAll
  - bug fix: Fixed Firefox issue where Delete would cause the display to "freeze"

Update 20 Jul 2022
  - bug fix. fixed problem with INPUT@5,19:A$ not returning a filename when file is empty
  - improved description field handling during edit
  - feature request: added secondary command 9. INPUT@5,9:A$ now returns a directory name of 'root'

update 15/07/2022
  - restored "MUSIC" call in FastGraphics ROM back to "MUZAKT" for compatibility with MaxiROM

Update 12/07/2022
  - bug fix: fixed problem with description field not getting padded when file name is generated
  - replaced hyphens to indicate empty character positions with space in filename display

Update 23/06/2022
  - added feature: added CLOSE command to Storage functions for compatibility. Some BASIC programs hung on encountering a CLOSE being sent to Storage over GPIB.

Update 28 Jan 2022

  - bug fix: File informaton was not getting displayed or saved during import of Flash drive files

Update 11 Jan 2022

  - bug fix: File information set by the user was not getting saved when a new file was added to storage.

Update 02 Dec 2021

  - feature request: added ability to edit description field
  - feature request: added support for dir command (secondary address 19)
  - miscellaneous bug fixes
  - removed support for 'Secret'

Update 05 Nov 2021:

  - export to '4924 Emulator Directory':
    - now renamed to '4050 Tape Emulator Files'
    - now exports all files to a single ZIP file

  - fixed bug where SAVE@5: appended to the previous file instead of overwriting it


Update 04 Nov 2021:

  - fixed some bugs reported by Monty McGraw including:
    - unclear functions of buttons: ClrALL renamed to DellALL. Delete renamed to DelFile. order of buttons changed to reflect a more logical grouping.
    - ClrALL (now DelALL) failed to delete the repository
    - import 4924 directory did not work
    - Clear did not clear file number or parameters to defaults


Update 02 Nov 2021:

  - the display has been updated by Monty McGraw to enable refresh graphics

  - implemented additional commands to read/write from web storage: PRINT, INPUT, WRITE, READ, BSAVE, BOLD

  - implemented ROM Expander (4051E01)

  - implemented ROM backpack right slot

  &nbsp;&nbsp;&nbsp;Known issues:

  - export to 4924 Emulator directory does not export all files. It will, however, export up to around 10 files at a time.


Update 16 Oct 2021:

  - implemented Dave R's fix for the random number generator


Update 15 Oct 2021:

  - added feature to allow import/export of files between the JavaScript emulator and Monty's Arduino 4924 emulator

  - added feature to allow the entire storage to be archived/imported as a single "tape" file archive"

  - updated Keyboard.js and re-mapped keyboard to use standard PC keys


Previous:

  - added feature to drag and drop files on to the storage window

  - added local storage features to enable programs to be stored in the browser and accessed like an external tape drive

---

<B>Additions by Monty McGraw:</B>

Update 02 Nov 2021

  - the display has been updated to enable refresh graphics

Previous:

 - Added new "UNIVERSAL" HTML emulator file that is compatible with all 4050 Control Characters

I plan to begin converting ALL my older program captures - to a universal format compatible with the universal emulator HTLM file      AND the 4051, 4052, 4052A and 4054A computers.

For the older programs from the Programs directory, and my older programs on the 4050 Program repository that have Character, backspace, underline - execute the older HTML file




TekDisplay.js


function TekDisplay(hw, canvas) {
	
	var canvasctx = canvas.getContext("2d");
	var width = canvas.width;
	var height = canvas.height;

// console.log(canvas.height + "," + canvas.width);

	var X_DA = 0;
	var Y_DA = 0;
	
	const pixel_erase_inten  = 255;
	const pixel_store_inten  = 240;
	const pixel_cursor_inten = 200;
	
	// Adjust position of graphics relative to canvas
	const screen_x_offset = 16;
	const screen_y_offset = 10;
	
	
    // ********************
    // ***              ***
    // ***  STD VIDEO.  ***
    // ***              ***
    // ********************

    FGColourIndex = 0x2;
    BGColourIndex = 0x0;

    // Colours are indexed [0..7] and then [Red,Green,Blue] (where Red, Green and Blue are in the range 0..255).
    //
    ColourTable = [
      [   0,   0,   0 ], // [0] Black.
      [   0,   0, 255 ], // [1] Blue.
      [   0, 255,   0 ], // [2] Green.
      [ 255,   0,   0 ], // [3] Red.
      [   0, 255, 255 ], // [4] Cyan.
      [ 255,   0, 255 ], // [5] Magenta.
      [ 255, 255,   0 ], // [6] Yellow.
      [ 255, 255, 255 ], // [7] White.
    ];
    
	canvasctx.fillStyle = "rgb("+ColourTable[0][0]+","+ColourTable[0][1]+","+ColourTable[0][2]+")";
	canvasctx.fillRect( 0, 0, width, height );
    
    // ****************
    // ***          ***
    // ***  VECTOR  ***
    // ***          ***
    // ****************

	function VECTOR( x0, y0, x1, y1 ) {
		var dx = Math.abs( x1 - x0 );
		var dy = Math.abs( y1 - y0 );
		
		var sx = -1;
		var sy = -1;
		
		if( x0 < x1 ) sx = 1;
		if( y0 < y1 ) sy = 1;
		
		var err = dx - dy;
		
		do {
		
			if (VEN_1 == 1) {
			
			// double vector pixel width and height for normal vectors
/*
			setPixel( x0+2, y0+2, 'VECTOR' );
			setPixel( x0+1, y0+2, 'VECTOR' );
			setPixel( x0+2, y0+1, 'VECTOR' );
			setPixel( x0+1, y0+1, 'VECTOR' );
*/
                setPixel(x0, y0, 'VECTOR', 2 );
			} else {
			// single vector pixel width and height for refresh vectors
//			setPixel( x0+1, y0+1, 'RVECTOR' );
                setPixel(x0+1, y0+1, 'RVECTOR', 1 );
			}
			
			if( (x0 == x1) && (y0 == y1) ) break;
			
			var e2 = 2 * err;
			
			if( e2 > -dy ) {
			
				err -= dy;
				
				x0 += sx;
				
			} // End if.
			
			if( e2 < dx ) {
			
				err += dx;
				
				y0 += sy;
				
			} // End if.
			
		} while( true );
			
	} // End of function VECTOR.


    // xpos, ypos = tektronix display co-ordinates
    // Added psize to indicate the pixel size (values 1=1x1 or 2=2x2)
    // to solve a FF rendering problem when 'Delete' key is pressed
	function setPixel( xpos, ypos, type, pxsize ) {

        if (pxsize < 1) pxsize = 1;

//		my = height - ypos; // Convert to 'canvas' coordinates from Tektronix coordinates.
        cypos = height - (ypos+1);  // Both have Zero base. Offset of 1 down required when converting.
  
		switch( type ) {
			case 'ERASE' :
		 		setPixelRGB( xpos, cypos, 0, 0, 0, pxsize ); // BLACK
				break;
			case 'SOT' : // cursor refresh dot
				// Do not replace pixel if it already has been stored!
				if(canvasctx.getImageData(xpos, cypos, 1, 1).data[1] != pixel_store_inten) {
				    setPixelRGB( xpos, cypos, 0, pixel_cursor_inten, 0, pxsize ); // DARK GREEN
				}
				break;
			case 'RVECTOR' : // vector refresh dot  -- added by mcm for refresh vector support
				// Do not replace pixel if it already has been stored!
				if(canvasctx.getImageData(xpos, my, 1, 1).data[1] != pixel_store_inten) {
				    setPixelRGB( xpos, cypos, 0, pixel_cursor_inten, 0, pxsize ); // DARK GREEN
				}
				break;
			case 'ADOT' : // stored character dot
				setPixelRGB( xpos, cypos, 0, pixel_store_inten, 0, pxsize ); // BRIGHT GREEN
				break;
			case 'VECTOR' :
				setPixelRGB( xpos, cypos, 0, pixel_store_inten, 0, pxsize ); // BRIGHT GREEN
				break;
			default :
				break;
		}					
	}


    // Draws a "pixel"
    // cxpos,cypos = canvas co-ordinates
	function setPixelRGB( cxpos, cypos, r, g, b, pxsize ) {
		canvasctx.fillStyle = "rgb("+r+","+g+","+b+")";
		canvasctx.fillRect( cxpos, cypos, pxsize, pxsize );
	}
    
   
    // ****************
    // ***          ***
    // ***  SCREEN  ***
    // ***          ***
    // ****************

	var adotpending = false;
	
	this.SCREEN = function( type ) {
        //console.log("TekDisplay SCREEN");

        DISP_INFO = hw.getDisplayControl();

		X_CHAR = DISP_INFO[0];
		Y_CHAR = DISP_INFO[1];
		
		VECTOR_0 = DISP_INFO[2];
		VEN_1    = DISP_INFO[3];
		
		new_X = DISP_INFO[4] + screen_x_offset;
		new_Y = DISP_INFO[5] + screen_y_offset;

		if ( type == 'BUFCLK' ) {
		
			var old_X = X_DA;
			var old_Y = Y_DA;

			// moved VEN_1 test to function VECTOR to enable persistent AND refresh vectors
			if( (VECTOR_0 == 0) )
				VECTOR( old_X, old_Y, new_X, new_Y );
		    
			// DER 9th August 2014 - Add debug code for drawing vectors.
			if( (VEN_1 == 1) && (VECTOR_0 == 0) && 0 ) {
			
				console.log( 'DER: TEKTRONIX4051.js - $$$' );
				console.log( ' DRAW VECTOR ' ); 
				console.log( ' old_X = '     ); console.log( old_X );
				console.log( ' old_Y = '     ); console.log( old_Y );
				console.log( ' new_X = '     ); console.log( new_X );
				console.log( ' new_Y = '     ); console.log( new_Y );
				console.log( ' VECTOR-0 = '  ); console.log( VECTOR_0 );
				console.log( ' VEN-1 = '     ); console.log( VEN_1 );
				
			} // End if.
			
			X_DA = new_X;
			Y_DA = new_Y;
		
		} // End if bufclk.
	
		// Check for cursor dot.
		if( (type == 'SOT') && (VECTOR_0 == 1) && (VEN_1 == 0) ) {
			if( adotpending ) {
//console.log("X_DA: " + X_DA + "  Y_DA: " + Y_DA + "  X_CHAR: " + X_CHAR + "  Y_CHAR: " + Y_CHAR);

				// setPixel( X_DA + X_CHAR, Y_DA + Y_CHAR, 'ADOT');
				// double width and double height of character pixels
				// 4 setPixel calls to interpolate between pixels for continuous font appearance
/*
				setPixel( X_DA + ((2*X_CHAR)+2), Y_DA + (2*Y_CHAR), 'ADOT', 1 );
				setPixel( X_DA + ((2*X_CHAR)+1), Y_DA + (2*Y_CHAR), 'ADOT', 1 );
    			setPixel( X_DA + ((2*X_CHAR)+2), Y_DA + ((2*Y_CHAR)-1), 'ADOT', 1 );
				setPixel( X_DA + ((2*X_CHAR)+1), Y_DA + ((2*Y_CHAR)-1), 'ADOT', 1 );
*/
                setPixel ( X_DA + (2*X_CHAR), Y_DA + (2*Y_CHAR), 'ADOT', 2 );
			} else {
			    // setPixel( X_DA + X_CHAR, Y_DA + Y_CHAR, 'SOT');
				// double width and double height of cursor pixels
				// 2 setPixel calls instead of 4 because it's pixelated like on actual machine
//				setPixel( X_DA + 2*X_CHAR+1, Y_DA + 2*Y_CHAR, 'SOT', 1 );
//				setPixel( X_DA + 2*X_CHAR+2, Y_DA + 2*Y_CHAR-1, 'SOT', 1 );

				setPixel( X_DA + (2*X_CHAR), Y_DA + (2*Y_CHAR), 'SOT', 1 );
				setPixel( X_DA + ((2*X_CHAR)+1), Y_DA + ((2*Y_CHAR)+1), 'SOT', 1 );

			}
			adotpending = false;
		} // End if sot.
		
		// Check for alphanumeric dot.
		if( (type == 'ADOT') && (VECTOR_0 == 1) && (VEN_1 == 0) ) {
			adotpending = true;
		}
		
		// DER - Debug code...
		if( 0 ) {
			console.log( 'DER: TEKTRONIX4051.js - $$$' );
			console.log( ' TYPE = '     ); console.log( type   );
			console.log( ' X_DA = '     ); console.log( X_DA   );
			console.log( ' Y_DA = '     ); console.log( Y_DA   );
			console.log( ' X_CHAR = '   ); console.log( X_CHAR );
			console.log( ' Y_CHAR = '   ); console.log( Y_CHAR );
			console.log( ' VECTOR-0 = ' ); console.log( VECTOR_0 );
			console.log( ' VEN-1 = '    ); console.log( VEN_1 );
		} // End if

	} // End of function SCREEN.


    // ***************
    // ***         ***
    // ***  ERASE  ***
    // ***         ***
    // ***************

	this.ERASE = function() {
		//console.log("TekDisplay: Erase screen");
        var imgd = canvasctx.getImageData(0, 0, width, height);
        var buffer = imgd.data.buffer;
        var buf8 = new Uint8ClampedArray(buffer);
        var data = new Uint32Array(buffer);
	
	    // Fill the screen bright green
	    for(i = 0; i < data.length; i++) {
           // 31:24 = alpha
           // 23:16 = blue
           // 15:8  = green
           //  7:0  = red
           data[i] = 0xFF00FF00;
        }
        imgd.data.set(buf8);
        canvasctx.putImageData(imgd, 0, 0);
        
        // Blank the screen
        for(i = 0; i < data.length; i++) {
           // 31:24 = alpha
           // 23:16 = blue
           // 15:8  = green
           //  7:0  = red
           data[i] = 0xFF000000;
        }
        imgd.data.set(buf8);
        
        // The DVST resets about 500 ms after the screen flash
        setTimeout(function(){
            canvasctx.putImageData(imgd, 0, 0);
            hw.displayReady();
            }, 500);	
	}

    
	this.dvst_emulate = function() {
	    var imgd = canvasctx.getImageData(0, 0, width, height);
        var buffer = imgd.data.buffer;
        var buf8 = new Uint8ClampedArray(buffer);
        var data = new Uint32Array(buffer);
	
	    for(i = 0; i < data.length; i++) {
           // 31:24 = alpha
           // 23:16 = blue
           // 15:8  = green
           //  7:0  = red
            var pixel = (data[i] >> 8) & 0xFF;
            
            // If the pixel is bright enough for the DVST to store it
            // then it will stay that way. If the pixel brightness
            // is below the threshold for the DVST to store the pixel
            // then its brightness will decay to zero.
            if(pixel_store_inten > pixel) {
               data[i] = 0xFF000000; // Blank the pixel
            }
        }
        imgd.data.set(buf8);
        canvasctx.putImageData(imgd, 0, 0); 
	}

	
	this.COPY = function() {
        var link = document.createElement('a');
        link.download = "screen.png";
        link.href = canvas.toDataURL("image/png").replace("image/png", "image/octet-stream");;
        link.click();
	
	}

}


TekKeyboard.js
function TekKeyboard(hw, windowobj) {

    // *****************************
    // ***                       ***
    // ***  HandleKeyboardEvent  ***
    // ***                       ***
    // *****************************

	var capsLockSet = false;

    function HandleKeyboardEvent( k, bool, scancode, e ) {

		var i = 0;


		//console.log("Key pressed: " + k + "  ASCII code: " + k.charCodeAt());


		//@@@ this.print('Key = '); this.printHex4( i ); this.print(' bool = '); this.print( bool ); this.print(' PC=0x'); this.printHex4( this.last_PC ); this.println('');

		// Deal with Shift, Control and CapsLock 
		switch (k) {
			case 'Shift' :
				hw.KBD_SHIFT_0 = bool ? 0 : 1;	// Shift key down = logical '0' (negative logic sense).
												// Shift key up   = logical '1'. 
				//e.preventDefault();
				k = '';
				return;
				break;
			case 'Control' :
				hw.KBD_CTRL_0 = bool ? 0 : 1;	// Control key down = logical '0' (negative logic sense).
												// Control key up   = logical '1'.
				k = '';
				return;
				break;

			case 'CapsLock' :
				if (bool) {
					capsLockSet = !capsLockSet;			// On capslock down toggle state of capslock
					hw.KBD_TTY_0 = capsLockSet ? 0 : 1;	// Caps lock key down = logical '0' (negative logic sense).
				}										  // Caps lock key up   = logical '1'. 
				k = '';
				return;
				break;
		}

		// console.log("KBD_SHIFT: " + hw.KBD_SHIFT_0);
		// console.log("K length: " + k.length);

		// Decide which key to use (some keys require translation to Tek codes)
		if ( k != '' ) {
			if ( k.length > 1 ) {
				// Look up the key ASCII character
				i = keyLookup(k);
			}else{
				// Check for Tek shift mapping
				hw.KBD_SHIFT_0 = tekShift(k.charCodeAt());
				// Look up the key string
				i = asciiLookup[k.charCodeAt()];
			}
		}else{
			if (scancode>0) {
				// Pass keyboard scan code directly
				i = scancode;
			}else{
				i = 0;
				return;
			}
		}

		// console.log("Key to Tek: " + i);
		// console.log("Shift status: " + hw.KBD_SHIFT_0);	// Down = 0, UP = 1;

		
		// Check for a key press or key release.
		if( bool ) {
		    hw.keyboardPress(i);
		} else {
		    hw.keyboardRelease();
		}

		// console.log("Done.");

		//keyboardInterrupt();
		
      	e.returnValue = false;
      	
		return false;
		
    } // End of function HandleKeyboardEvent.
    
    
    // *********************
    // ***               ***
    // ***  HandleEvent  ***
    // ***               ***
    // *********************

	// Hardware key events
    function handleEvent( e ) {
        var storage = document.getElementById('storage');
		// If storage dialog is displayed then ignore and return
        if (storage) {
            if (storage.style.display=="block") return;
        }
		// Did we get an emulated keypress scan code
		if (e.detail.scancode) {
			return HandleKeyboardEvent('', e.detail.type=='keydown', e.detail.scancode, e);
		}else{		
			// Otherwise execute keyboard handler
        	return HandleKeyboardEvent(e.key, e.type=='keydown', 0, e );
		}
    }

/*    
	// Emulated key events
	function handleEkeyEvent ( e ) {
        var storage = document.getElementById('storage');
        if (storage) {
            if (storage.style.display=="block") return;
        }
		
	}
*/

	// ***************
    // ***         ***
    // ***  FcnKey ***
    // ***         ***
    // ***************
/*
	this.FcnKey = function(code, press) {
		// code - key code for keyboard key
		// press   - true for keydown, false for key up
		var e = {keyCode:code, type:press ? "keydown" : "keyup", returnValue: false};
		return handleEvent(e);
	}
*/

	// Most characters can be looked up in a table
	var asciiLookup = [
	//  Tek:-	   PC:-
		0x00,	// 000 [00] Null
		0x00,	// 001 [01] SOH (Start of heading)
		0x00,	// 002 [02] STX (Start of text)
		0x00,	// 003 [03] ETX (End of text)
		0x00,	// 004 [04] EOT (End of transmission)
		0x00,	// 005 [05] ENQ (Enquiry)
		0x00,	// 006 [06] ACK (Acknowledge)
		0x00,	// 007 [07] BEL (Bell)
		0x08,	// 008 [08] BS  (Backspace)
		0x09,	// 009 [09] TAB (Horizontal tab)
		0x0A,	// 010 [0A] LF  (Line feed)
		0x00,	// 011 [0B] VT  (Vertical tab)
		0x00,	// 012 [0C] FF  (Form feed)
		0x0D,	// 013 [0D] CR  (Carriage return)
		0x00,	// 014 [0E] SO  (Shift out)
		0x00,	// 015 [0F] SI  (Shift in)
		0x10,	// 016 [10] DLE (Data link escape)
		0x00,	// 017 [11] DC1 (Device control 1)
		0x00,	// 018 [12] DC1 (Device control 2)
		0x00,	// 019 [13] DC1 (Device control 3)
		0x00,	// 020 [14] DC1 (Device control 4)
		0x00,	// 021 [15] NAK (Negative acknowledge)
		0x00,	// 022 [16] SYN (Synchronous idle)
		0x00,	// 023 [17] ETB (End of transmiaaion block)
		0x00,	// 024 [18] CAN (Cancel)
		0x00,	// 025 [19] EM  (End of medium)
		0x00,	// 026 [1A] SUB (Substitute)
		0x1B,	// 027 [1B] ESC (Escape)
		0x00,	// 028 [1C] FS  (File separator)
		0x00,	// 029 [1D] GS  (Group separator)
		0x00,	// 030 [1E] RS  (Record separator)
		0x00,	// 031 [1F] US  (Unit separator)
		0x10,	// 032 [20] Space
		0x31,	// 033 [21] !
		0x32,	// 034 [22] "
		0x33,	// 035 [23] # [':']
		0x34,	// 036 [24] $
		0x35,	// 037 [25] %
		0x36,	// 038 [26] &
		0x37,	// 039 [27] '      ' ['@'] 0x40 0x37
		0x38,	// 040 [28] (
		0x39,	// 041 [29] )
		0x2A,	// 042 [2A] *
		0x3B,	// 043 [2B] +
		0x3C,	// 044 [2C] ,
		0x3D,	// 045 [2D] -
		0x3E,	// 046 [2E] .
		0x3F,	// 047 [2F] /
		0x30,	// 048 [30] 0
		0x31,	// 049 [31] 1
		0x32,	// 050 [32] 2
		0x33,	// 051 [33] 3
		0x34,	// 052 [34] 4
		0x35,	// 053 [35] 5
		0x36,	// 054 [36] 6
		0x37,	// 055 [37] 7
		0x38,	// 056 [38] 8
		0x39,	// 057 [39] 9
		0x3A,	// 058 [3A] :
		0x3B,	// 059 [3B] ;
		0x3C,	// 060 [3C] <
		0x3D,	// 061 [3D] =
		0x3E,	// 062 [3E] >
		0x3F,	// 063 [3F] ?
		0x40,	// 064 [40] @
		0x41,	// 065 [41] A
		0x42,	// 066 [42] B
		0x43,	// 067 [43] C
		0x44,	// 068 [44] D
		0x45,	// 069 [45] E
		0x46,	// 070 [46] F
		0x47,	// 071 [47] G
		0x48,	// 072 [48] H
		0x49,	// 073 [49] I
		0x4A,	// 074 [4A] J
		0x4B,	// 075 [4B] K
		0x4C,	// 076 [4C] L
		0x4D,	// 077 [4D] M
		0x4E,	// 078 [4E] N
		0x4F,	// 079 [4F] O
		0x50,	// 080 [50] P
		0x51,	// 081 [51] Q
		0x52,	// 082 [52] R
		0x53,	// 083 [53] S
		0x54,	// 084 [54] T
		0x55,	// 085 [55] U
		0x56,	// 086 [56] V
		0x57,	// 087 [57] W
		0x58,	// 088 [58] X
		0x59,	// 089 [59] Y
		0x5A,	// 090 [5A] Z
		0x5B,	// 091 [5B] [
		0x5C,	// 092 [5C] Backslash
		0x5D,	// 093 [5D] ]
		0x5E,	// 094 [5E] ^
		0x5F,	// 095 [5F] _
		0x40,	// 096 [60] `
		0x41,	// 097 [61] a
		0x42,	// 098 [62] b
		0x43,	// 099 [63] c
		0x44,	// 100 [64] d
		0x45,	// 101 [65] e
		0x46,	// 102 [66] f
		0x47,	// 103 [67] g
		0x48,	// 104 [68] h
		0x49,	// 105 [69] i
		0x4A,	// 106 [6A] j
		0x4B,	// 107 [6B] k
		0x4C,	// 108 [6C] l
		0x4D,	// 109 [6D] m
		0x4E,	// 110 [6E] n
		0x4F,	// 111 [6F] o
		0x50,	// 112 [70] p
		0x51,	// 113 [71] q
		0x52,	// 114 [72] r
		0x53,	// 115 [73] s
		0x54,	// 116 [74] t
		0x55,	// 117 [75] u
		0x56,	// 118 [76] v
		0x57,	// 119 [77] w
		0x58,	// 120 [78] x
		0x59,	// 121 [79] y
		0x5A,	// 122 [7A] z
		0x5B,	// 123 [7B] {
		0x5C,	// 124 [7C] |
		0x5D,	// 125 [7D] }
		0x5E,	// 126 [7E] ~
		0x7F	// 127 [7F] Delete (numeric keypad)

	];


	// Special keys generate strings in the 'key' parameter 
	function keyLookup ( key ) {

		//console.log("Looking up " + key);

		switch (key) {
			// Control keys
			case "Alt"         : return 0x00; break;
			case "AltGraph"    : return 0x00; break;
			case "ArrowDown"   : return 0x00; break;
			case "ArrowLeft"   : return 0x00; break;
			case "ArrowRight"  : return 0x00; break;
			case "ArrowUp"     : return 0x00; break;
			case "Backspace"   : return 0x08; break;	// Back space
			case "Control"     : return 0x11; break;
			case "Capslock"    : return 0x14; break;
			case "ContextMenu" : return 0x00; break;
			case "Delete"      : return 0x5F; break;	// Rubout
			case "End"         : return 0x6B; break;	// BREAK
			case "Enter"       : return 0x0D; break;
			case "Escape"      : return 0x1B; break;
			case "F1"   	   : return 0x60; break; 
			case "F2"   	   : return 0x61; break; 
			case "F3"   	   : return 0x62; break; 
			case "F4"   	   : return 0x63; break; 
			case "F5"   	   : return 0x64; break; 
			case "F6"   	   : return 0x65; break; 
			case "F7"   	   : return 0x66; break;
			case "F8"   	   : return 0x67; break; 
			case "F9"   	   : return 0x68; break; 
			case "F10"   	   : return 0x69; break; 
			case "F11"   	   : return 0x00; break; 
			case "F12"   	   : return 0x00; break;
			case "F13"   	   : return 0x6A; break;	// [Mac] PAGE 
			case "F14"   	   : return 0x6B; break;	// [Mac] BREAK
			case "F15"   	   : return 0x73; break;	// [Mac] CLEAR 
			case "F16"   	   : return 0x75; break;	// [Mac] AUTO NUM 
			case "F17"   	   : return 0x76; break;	// [Mac] STEP PROG 
			case "F18"   	   : return 0x00; break;	// [Mac]
			case "F19"   	   : return 0x00; break;	// [Mac]
			case "Home"        : return 0x6A; break; 	// PAGE£
			case "Insert"      : return 0x73; break; 	// CLEAR
			case "NumLock"     : return 0x70; break; 
			case "PageDown"    : return 0x00; break;
			case "PageUp"      : return 0x75; break;	// AUTO NUM
			case "Pause"       : return 0x76; break;	// STEP PROG
			case "ScrollLock"  : return 0x00; break;
			case "Shift"       : return 0x10; break;
			case "Tab"         : return 0x09; break;
		}
	}

	// Some characters require the status of shift to be reversed
	function tekShift(key) {

		switch (key) {
			case 0x23 : return !hw.KBD_SHIFT_0; break; // 7 => #
			case 0x27 : return !hw.KBD_SHIFT_0; break; // 7 => '	
//			case 0x2A : return !hw.KBD_SHIFT_0; break; // *		
			case 0x3A : return !hw.KBD_SHIFT_0; break; // * => :
			case 0x3D : return !hw.KBD_SHIFT_0; break; // - => =
			case 0x40 : return !hw.KBD_SHIFT_0; break; // ` => @
			case 0x5E : return !hw.KBD_SHIFT_0; break; // ~ => ^
			default : 	return hw.KBD_SHIFT_0;
		}	
	};

	// Hardware key events
    windowobj.onkeydown = handleEvent;
    windowobj.onkeyup   = handleEvent;

	// Emulated key events
	windowobj.addEventListener('kdbePress', function(ev) {
		handleEvent(ev);
	});
	windowobj.addEventListener('kdbeRelease', function(ev) {
		handleEvent(ev);
	});

}

